### ISSUE TITLE
DocumentClient.batchWrite return circular structure

### ISSUE THREAD

**Issue by seoker** [aws/aws-sdk-js#1474]
While I was testing `DocumentClient.batchWrite` method, I found the result has changed with circular structure so that I cannot print the result in JSON.

The results is ok with `aws-sdk-js v2.43.0`, but not with `v2.44.0`.

I'm using `restify v4.3.0`, not sure if it's related.

```javascript
const params = {
  RequestItems: requestItems
};

docClient.batchWrite(params).promise().then((response) => {
  console.log('response:', response);
});
```

Response will be like this:
```json
    UnprocessedItems={},
    $response: {
      "request": {
        "service": {
          "config": {
            "credentials": {
              "expired": false,
              "expireTime": null,
              "accessKeyId": "{my-access-key-id}",
              "profile": "{my-profile}",
              "disableAssumeRole": true
            },
            "credentialProvider": {
              "providers": [
                null,
                null,
                null,
                null
              ]
            },
            "region": "us-west-2",
            "logger": null,
            "apiVersions": {},
            "apiVersion": null,
            "endpoint": "dynamodb.us-west-2.amazonaws.com",
            "httpOptions": {
              "timeout": 120000
            },
            "maxRedirects": 10,
            "paramValidation": true,
            "sslEnabled": true,
            "s3ForcePathStyle": false,
            "s3BucketEndpoint": false,
            "s3DisableBodySigning": true,
            "computeChecksums": true,
            "convertResponseTypes": true,
            "correctClockSkew": false,
            "customUserAgent": null,
            "dynamoDbCrc32": true,
            "systemClockOffset": 0,
            "signatureVersion": "v4",
            "signatureCache": true,
            "retryDelayOptions": {},
            "useAccelerateEndpoint": false
          },
          "isGlobalEndpoint": false,
          "endpoint": {
            "protocol": "https:",
            "host": "dynamodb.us-west-2.amazonaws.com",
            "port": 443,
            "hostname": "dynamodb.us-west-2.amazonaws.com",
            "pathname": "/",
            "path": "/",
            "href": "https://dynamodb.us-west-2.amazonaws.com/"
          },
          "_clientId": 9
        },
        "operation": "batchWriteItem",
        "params": {
          "RequestItems": {
            "user-program": [
              {
                "PutRequest": {
                  "Item": {
                    ... // my data in valid format
                  }
                }
              },
              {
                "PutRequest": {
                  "Item": {
                    ... // my data in valid format
                  }
                }
              },
              {
                "PutRequest": {
                  "Item": {
                    ... // my data in valid format
                  }
                }
              }
            ]
          }
        },
        "httpRequest": {
          "method": "POST",
          "path": "/",
          "headers": {
            "User-Agent": "aws-sdk-nodejs/2.44.0 darwin/v7.6.0 promise",
            "Content-Type": "application/x-amz-json-1.0",
            "X-Amz-Target": "DynamoDB_20120810.BatchWriteItem",
            "X-Amz-Content-Sha256": "4e69510bc0695ca4b25e1422c7028a0987713b34ee499d01430c8e7d567eab03",
            "Content-Length": 857,
            "Host": "dynamodb.us-west-2.amazonaws.com",
            "X-Amz-Date": "20170421T051010Z",
            "Authorization": "{authorization-token}"
          },
          "body": "{\"RequestItems\":{\"user-program\":[{\"PutRequest\":{\"Item\":{\"userId\":{\"S\":\"1124649538729634596\"},\"createdAt\":{\"S\":\"2017-04-21T05:10:10Z\"},\"updatedAt\":{\"S\":\"2017-04-21T05:10:10Z\"},\"status\":{\"S\":\"HL00BW-S10\"},\"code\":{\"S\":\"HL00BW\"},\"source\":{\"S\":\"user\"},\"programId\":{\"S\":\"7a83f3d3-56e6-41a9-aaf4-8b8aef07f7cb\"}}}},{\"PutRequest\":{\"Item\":{\"userId\":{\"S\":\"1124649538729634596\"},\"createdAt\":{\"S\":\"2017-04-21T05:10:10Z\"},\"updatedAt\":{\"S\":\"2017-04-21T05:10:10Z\"},\"status\":{\"S\":\"HL00CP-S10\"},\"code\":{\"S\":\"HL00CP\"},\"source\":{\"S\":\"user\"},\"programId\":{\"S\":\"91338330-920a-48e3-8c98-673707390361\"}}}},{\"PutRequest\":{\"Item\":{\"userId\":{\"S\":\"1124649538729634596\"},\"createdAt\":{\"S\":\"2017-04-21T05:10:10Z\"},\"updatedAt\":{\"S\":\"2017-04-21T05:10:10Z\"},\"status\":{\"S\":\"HL00CC-S10\"},\"code\":{\"S\":\"HL00CC\"},\"source\":{\"S\":\"user\"},\"programId\":{\"S\":\"fa7e44e8-d429-47db-a723-fc8348bf2f90\"}}}}]}}",
          "endpoint": {
            "protocol": "https:",
            "host": "dynamodb.us-west-2.amazonaws.com",
            "port": 443,
            "hostname": "dynamodb.us-west-2.amazonaws.com",
            "pathname": "/",
            "path": "/",
            "href": "https://dynamodb.us-west-2.amazonaws.com/"
          },
          "region": "us-west-2",
          "_userAgent": "aws-sdk-nodejs/2.44.0 darwin/v7.6.0 promise",
          "stream": {
            "domain": null,
            "_events": {},
            "_eventsCount": 5,
            "output": [],
            "outputEncodings": [],
            "outputCallbacks": [],
            "outputSize": 0,
            "writable": true,
            "_last": false,
            "upgrading": false,
            "chunkedEncoding": false,
            "shouldKeepAlive": true,
            "useChunkedEncodingByDefault": true,
            "sendDate": false,
            "_removedHeader": {
              "content-length": false
            },
            "_contentLength": 857,
            "_hasBody": true,
            "_trailer": "",
            "finished": true,
            "_headerSent": true,
            "socket": {
              "_tlsOptions": {
                "pipe": null,
                "secureContext": {
                  "context": {},
                  "singleUse": true
                },
                "isServer": false,
                "requestCert": true,
                "rejectUnauthorized": true,
                "session": {
                  "type": "Buffer",
                  "data": [
                    48,
                    130,
                    6,
                    122,
                    2,
                    1,
                    1,
                    2,
                    2,
                    3,
                    3,
                    4,
                    2,
                    192,
                    ... // and many others
                  ]
                }
              },
              "_secureEstablished": true,
              "_securePending": false,
              "_newSessionPending": false,
              "_controlReleased": true,
              "_SNICallback": null,
              "servername": null,
              "npnProtocol": false,
              "alpnProtocol": false,
              "authorized": true,
              "authorizationError": null,
              "encrypted": true,
              "_events": {
                "close": [
                  null,
                  null,
                  null
                ],
                "end": [
                  null
                ]
              },
              "_eventsCount": 9,
              "connecting": false,
              "_hadError": false,
              "_handle": null,
              "_parent": null,
              "_host": "dynamodb.us-west-2.amazonaws.com",
              "_readableState": {
                "objectMode": false,
                "highWaterMark": 16384,
                "buffer": {
                  "head": null,
                  "tail": null,
                  "length": 0
                },
                "length": 0,
                "pipes": null,
                "pipesCount": 0,
                "flowing": true,
                "ended": false,
                "endEmitted": false,
                "reading": true,
                "sync": false,
                "needReadable": true,
                "emittedReadable": false,
                "readableListening": false,
                "resumeScheduled": false,
                "defaultEncoding": "utf8",
                "ranOut": false,
                "awaitDrain": 0,
                "readingMore": false,
                "decoder": null,
                "encoding": null
              },
              "readable": false,
              "domain": null,
              "_writableState": {
                "objectMode": false,
                "highWaterMark": 16384,
                "needDrain": false,
                "ending": false,
                "ended": false,
                "finished": false,
                "decodeStrings": false,
                "defaultEncoding": "utf8",
                "length": 0,
                "writing": false,
                "corked": 0,
                "sync": false,
                "bufferProcessing": false,
                "writecb": null,
                "writelen": 0,
                "bufferedRequest": null,
                "lastBufferedRequest": null,
                "pendingcb": 0,
                "prefinished": false,
                "errorEmitted": false,
                "bufferedRequestCount": 0,
                "corkedRequestsFree": {
                  "next": {
                    "next": null,
                    "entry": null
                  },
                  "entry": null
                }
              },
              "writable": false,
              "allowHalfOpen": false,
              "destroyed": true,
              "_bytesDispatched": 1473,
              "_sockname": null,
              "_pendingData": null,
              "_pendingEncoding": "",
              "_server": null,
              "ssl": null,
              "_requestCert": true,
              "_rejectUnauthorized": true,
              "parser": null,
              "_httpMessage": "[Circular]",
              "_idleTimeout": -1,
              "_idleNext": null,
              "_idlePrev": null,
              "_idleStart": 7482,
              "_consuming": true
            },
            "connection": "[Circular]",
            ... // many others that comes with circular structure
          }
        },
        "startTime": "2017-04-21T05:10:10.231Z",
        "response": "[Circular]",
        "_asm": {
          "currentState": "complete",
          "states": {
            "validate": {
              "accept": "build",
              "fail": "error"
            },
            "build": {
              "accept": "afterBuild",
              "fail": "restart"
            },
            "afterBuild": {
              "accept": "sign",
              "fail": "restart"
            },
            "sign": {
              "accept": "send",
              "fail": "retry"
            },
            "retry": {
              "accept": "afterRetry",
              "fail": "afterRetry"
            },
            "afterRetry": {
              "accept": "sign",
              "fail": "error"
            },
            "send": {
              "accept": "validateResponse",
              "fail": "retry"
            },
            "validateResponse": {
              "accept": "extractData",
              "fail": "extractError"
            },
            "extractError": {
              "accept": "extractData",
              "fail": "retry"
            },
            "extractData": {
              "accept": "success",
              "fail": "retry"
            },
            "restart": {
              "accept": "build",
              "fail": "error"
            },
            "success": {
              "accept": "complete",
              "fail": "complete"
            },
            "error": {
              "accept": "complete",
              "fail": "complete"
            },
            "complete": {
              "accept": null,
              "fail": null
            }
          }
        },
        "_haltHandlersOnError": false,
        "_events": {
          "validate": [
            null,
            null,
            null,
            null,
            null
          ],
          "afterBuild": [
            null,
            null,
            null
          ],
          "restart": [
            null
          ],
          "sign": [
            null
          ],
          "validateResponse": [
            null
          ],
          "send": [
            null
          ],
          "httpHeaders": [
            null
          ],
          "httpData": [
            null
          ],
          "httpDone": [
            null
          ],
          "retry": [
            null,
            null,
            null,
            null,
            null,
            null
          ],
          "afterRetry": [
            null
          ],
          "build": [
            null
          ],
          "extractData": [
            null,
            null,
            null,
            null
          ],
          "extractError": [
            null,
            null
          ],
          "httpError": [
            null
          ],
          "complete": [
            null
          ]
        },
        "rawParams": {
          "RequestItems": {
            "user-program": [
              {
                "PutRequest": {
                  "Item": {
                    ... // my item in valid format
                  }
                }
              },
              {
                "PutRequest": {
                  "Item": {
                    ... // my item in valid format
                  }
                }
              },
              {
                "PutRequest": {
                  "Item": {
                    ... // my item in valid format
                  }
                }
              }
            ]
          }
        },
        "signedAt": "2017-04-21T05:10:10.234Z"
      },
      "data": {
        "UnprocessedItems": "[Circular]"
      },
      "error": null,
      "retryCount": 0,
      "redirectCount": 0,
      "httpResponse": {
        "statusCode": 200,
        "headers": "[Circular]",
        "body": {
          "type": "Buffer",
          "data": [
            123,
            34,
            85,
            110,
            112,
            114,
            111,
            99,
            101,
            115,
            115,
            101,
            100,
            73,
            116,
            101,
            109,
            115,
            34,
            58,
            123,
            125,
            125
          ]
        },
        "streaming": false,
        "stream": "[Circular]",
        "statusMessage": "OK"
      },
      "maxRetries": 10,
      "maxRedirects": 10,
      "requestId": "00CA1Q2AKU8GISTDOTDDTGNJRJVV4KQNSO5AEMVJF66Q9ASUAAJG"
    }
```

**Comment by chrisradek** [comment]
@seoker
2.44.0 introduced a new field on the data returned by an operation when using promises that caused this issue.
A change was pushed to master last night to address the circular reference issue. Are you able to test with the master branch? We will publish a new version of the SDK to npm today as well.

Thanks for reporting the issue and apologies for the trouble this caused.

**Comment by chrisradek** [comment]
This has been fixed and released in version 2.45.0 of the SDK. Thanks for reporting the issue! If you encounter any issues after upgrading, please let us know.
